### ISSUE TITLE
Data returned by apollo service query does not include __typename

### ISSUE THREAD

**Issue by n1ru4l** [bgentry/ember-apollo-client#34]
When your query contains the `__typename` field is is stripped out.
This makes working with interfaces and store updates very hard.

**Comment by bgentry** [comment]
This is likely caused by the way we're using `Ember.copy` to copy from the query result to an `Ember.Object` (`copy` ignores private fields like `__typename`, I think). Unfortunately it wouldn't be solved by #20 either. Is there a better way to copy so that such fields are preserved? Or do you have any ideas for how to bridge the gap between ApolloClient and Ember's object model?

**Comment by viniciussbs** [comment]
@n1ru4l I guess the problem is with `Ember.copy`, that [does not copy properties beginning with `__`](https://github.com/emberjs/ember.js/blob/v2.12.0/packages/ember-runtime/lib/copy.js#L48-L52 ). And [it's used in the service](https://github.com/bgentry/ember-apollo-client/blob/master/addon/services/apollo.js#L132) because [Apollo client returns a immutable object](https://dev-blog.apollodata.com/apollo-client-0-8-e8db5efc75fb#2658), but Ember expects an extensible object.

**Comment by viniciussbs** [comment]
@bgentry if the object to be copied implements the `Ember.Copyable` interface, then `Ember.copy` will delegate to the object's `copy()` method and return the result. We could return an `ApolloResultObject` instead of a `Ember.Object` and implement `Ember.Copyable`. Our implementation could be the same of `Ember.copy` but copying `__typename`. 

Or we could use `JSON.parse(JSON.stringify(data))` instead.

@n1ru4l Could you try to alias `__typename` to something like `typename`?

```graphql
query {
  myThing {
    otherThing
    typename: __typename
  }
}
```

**Comment by bgentry** [comment]
The extensibility/immutability is one part of it. The other issue is that Ember needs to know when the result has changed (when to rerender/recompute), which gets handled automatically by copying into an Ember.Object.

**Comment by bgentry** [comment]
@viniciussbs adding our own copy implementation seems like the most straightforward way to solve this as you said.

**Comment by viniciussbs** [comment]
@n1ru4l I've tested a query using `typename: __typename` and it worked. You can change the alias to what fits better to your case. And if this solves your problem, I think we could include this into the README as a caveat while we don't have a proper solution.

**Comment by viniciussbs** [comment]
[I've opened a issue on Ember's repo](https://github.com/emberjs/ember.js/issues/15431) because I think we should be able to tell Ember that some properties prefixed with `__` are not some of its internal properties.

**Comment by n1ru4l** [comment]
@viniciussbs Thank you for the workaround, I will try that tomorrow!

**Comment by n1ru4l** [comment]
Looks like this will not get fixed any time soon. The workaround described by @viniciussbs does indeed work, nevertheless I would like to have a own copy implementation instead of mapping trough all objects when doing a optimistic response or similar.

I could do a PR for this. 
@bgentry What is your opinion on usinge the [Copyable](https://github.com/emberjs/ember.js/blob/v2.12.0/packages/ember-runtime/lib/mixins/copyable.js#L28) Mixin which is marked as private?

**Comment by viniciussbs** [comment]
[There's a documentation for `Ember.Copyable`](https://www.emberjs.com/api/classes/Ember.Copyable.html). But we could also save a copy function on addon/utils and use it.

I suggest we create something that could be used inside apps. I would use it inside my `updateQuery` reducers used inside `fetchMore` for pagination. That's where I ended up using `JSON.parse(JSON.stringify(data))`.

Here is an exemple of my macro that creates reduce functions for my Relay-style paginations:

```graphql
import Ember from 'ember';

const { get, set } = Ember;

// HACK: `Ember.copy` ignores property `__typename` that Apollo relies on.
function copy(object) {
  return JSON.parse(JSON.stringify(object));
}

export default function pushEdges(connectionPath) {
  let edgesPath = `${connectionPath}.edges`;
  let cursorPath = `${connectionPath}.pageInfo.endCursor`;

  return (previousResult, { fetchMoreResult }) => {
    let newEdges = get(fetchMoreResult, edgesPath);
    let newCursor = get(fetchMoreResult, cursorPath);
    let newResult = copy(previousResult);

    if (newCursor) {
      set(newResult, cursorPath, newCursor);
    }

    get(newResult, edgesPath).pushObjects(newEdges);

    return newResult;
  };
}
```

**Comment by n1ru4l** [comment]
@viniciussbs There is a documentation for it however it is marked as private, so I asked @locks and he confirmed that private mixins should not be used outside of ember-core. Since we would have to utilize `Ember.Copyable` to implement a custom copy function that reuses the `Ember.copy` implementation I do not think this is a good solution unless `Ember.Copyable` becomes public. Therefore the `JSON.parse(JSON.stringify(object))` approach seems best to me unless there could be a more performant way to copy the object.

**Comment by bgentry** [comment]
Stringify + parse would be way too inefficient. I'd rather whitelist specific additional keys to be copied if necessary.

**Comment by n1ru4l** [comment]
@bgentry So would like to reimplement `Ember.copy` which relies on the marked as private `Ember.Copyable`?

**Comment by bgentry** [comment]
We can call `Ember.copy` and then manually copy `__typename`. I can't say I like it, but I like it a lot more than stringifying and parsing again 🤷🏼‍♂️

**Comment by viniciussbs** [comment]
I guess the data object returned by Apollo is frozen, so we can't "rename" `__typename`.

We can use the [default implementation](https://github.com/emberjs/ember.js/blob/v2.12.0/packages/ember-runtime/lib/copy.js#L5-L64) and change it to copy whitelisted keys as @bgentry has suggested. We can remove the [part that handles objects that implement `Ember.Copyable`](https://github.com/emberjs/ember.js/blob/v2.12.0/packages/ember-runtime/lib/copy.js#L18-L21) because it's not our case.

Keys that could be whitelisted: [meta fields](http://graphql.org/learn/queries/#meta-fields) and [introspection fields](http://graphql.org/learn/introspection/).

```javascript
// Original implementation

if (key.substring(0, 2) === '__') {
  continue;
}
```

```javascript
// ember-apollo-client implementation

if (_isEmberInternalKey(key) && !_isGraphQLInternalKey(key)) {
  continue;
}
```

```javascript
function _isEmberInternalKey(key) {
  return key.substring(0, 2) === '__';
}

function _isGraphQLKey(key) {
  let keys = [
    '__typename',
    // ...
  ];

  return keys.indexOf(key) !== -1;
}
```

**Comment by bgentry** [comment]
@n1ru4l this is fixed in #44, please take a look if you can

**Comment by n1ru4l** [comment]
@bgentry Project got trashed, I do not think that I will use this library anytime soon. Sorry :(
